### PR TITLE
RST Cloud: Threat Feed Integration

### PR DESCRIPTION
**Pull request by k1r10n**
Signed-off-by: k1r10n <ysergeev@rstcloud.net>

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: ....

## Description
Aggregated, filtered, and scored threat intelligence database available via API to check domains, URLs, and IP addresses

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [x] 5.0.0
- [x] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ x ] No

## Must have
- [ x ] Tests
- [ x ] Documentation

### CODE REVIEW

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+    markdown: str = ''
+    results: List[str] = []
+    num = 0
+    for ioc in iocs:
+        indicator = client.submit_indicator(ioc, description[num])
+        num += 1
+        if "status" in indicator.keys():
+            markdown += f"Indicator: {ioc} was submitted as a potential threat indicator to RST Cloud\n"
+        elif "error" in indicator.keys():
+            markdown += f"Indicator: {ioc} was not submitted successfully due to the following error: {indicator['error']}\n"
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py:377]: If a command fails, it should raise an exception (e.g. `DemistoException`), not fail gracefully so the playbook will stop on errors. Main will catch it and handle it.

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+                     'Description': f'\n{additional_info["Description"]}\n',
+                     'Tags': ', '.join(tag for tag in additional_info['Tags'])}
+            markdown += tableToMarkdown(f'RST Threat Feed URL Reputation for: {url}\n'
+                                        f'{RST_URL}uuid?id={indicator["id"]}', table, removeNull=True)
+            results.append(indicator)
+
+    return markdown, context, results
+
+
+@logger
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py:363]: Don't use logger
```suggestion
```

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+    Args:
+      client (Client): RST Threat Feed client.
+    Returns:
+      str: 'ok' if test passed, anything else will fail the test.
+    """
+
+    return 'ok' if client.get_indicator('1.1.1.1') else 'Connection failed.'
+
+
+@logger
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py:115]: Don't use logger
```suggestion
```

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+                     'Description': f'\n{additional_info["Description"]}\n',
+                     'Tags': ', '.join(tag for tag in additional_info['Tags'])}
+            markdown += tableToMarkdown(f'RST Threat Feed IP Reputation for: {indicator["ioc_value"]}\n'
+                                        f'{RST_URL}uuid?id={indicator["id"]}', table, removeNull=True)
+            results.append(indicator)
+
+    return markdown, context, results
+
+
+@logger
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py:206]: Don't use logger
```suggestion
```

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+                     'Description': f'\n{additional_info["Description"]}\n',
+                     'Tags': ', '.join(tag for tag in additional_info['Tags'])}
+            markdown += tableToMarkdown(f'RST Threat Feed Domain Reputation for: {indicator["ioc_value"]}\n'
+                                        f'{RST_URL}uuid?id={indicator["id"]}', table, removeNull=True)
+            results.append(indicator)
+
+    return markdown, context, results
+
+
+@logger
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py:288]: Don't use logger
```suggestion
```

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+        indicator = client.submit_indicator(ioc, description[num])
+        num += 1
+        if "status" in indicator.keys():
+            markdown += f"Indicator: {ioc} was submitted as a potential threat indicator to RST Cloud\n"
+        elif "error" in indicator.keys():
+            markdown += f"Indicator: {ioc} was not submitted successfully due to the following error: {indicator['error']}\n"
+    return markdown, context, results
+
+
+@logger
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py:381]: Don't use logger
```suggestion
```

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+RST_URL = 'https://rstcloud.net/'
+DEFAULT_THRESHOLD = 50
+DEFAULT_EXPIRATION = 180
+DBOT_SCORE_KEY = 'DBotScore(val.Indicator == obj.Indicator && val.Vendor == obj.Vendor)'
+
+''' CLIENT CLASS '''
+
+
+class Client(BaseClient):
+    def __init__(self, apikey, api_url='https://api.rstcloud.net/v1', verify=False, proxy=False):
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py:20]: Please add type hints to the class methods.

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.yml [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+    - contextPath: RST.Domain.FalsePositive
+      description: true if it is likely a False Positive
+      type: String
+    - contextPath: RST.Domain.FalsePositiveDesc
+      description: Description why we think it may be a False Positive
+      type: String
+  - arguments:
+    - default: true
+      description: List of URLs
+      isArray: false
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.yml:312]:
```suggestion
      isArray: true
```

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.yml [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+    - contextPath: RST.URL.FalsePositive
+      description: true if it is likely a False Positive
+      type: String
+    - contextPath: RST.URL.FalsePositiveDesc
+      description: Description why we think it may be a False Positive
+      type: String
+  - arguments:
+    - default: true
+      description: List of IOCs (URLs, domains or IPs)
+      isArray: false
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.yml:408]:
```suggestion
      isArray: true
```

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.yml [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+      required: false
+      secret: false
+    deprecated: false
+    description: Submits an indicator to RST Threat Feed
+    execution: false
+    name: rst-threat-feed-submit
+  - arguments:
+    - default: true
+      description: List of IOCs (URLs, domains or IPs)
+      isArray: false
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.yml:425]:
```suggestion
      isArray: true
```

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/README.md [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+Also, the following contact details can be used:
+- **URL**: [https://www.rstcloud.net/contact](https://www.rstcloud.net/contact)
+
+Each indicator is ranked from 0 to 100. Indicators are being collected from multiple sources and are cross-verified using multiple criteria. 
+Please check indicator tags and malware family field. An indicator may describe you a known malware or a scanning host. Threfore, different actions may be required based on the context.
+
+Usually, to exlude False Positives use the following thresholds:
+ - for IPs you start with score=50
+ - for Domains you start with score=40
+ - for URLs you start with score=30
```
Comment (DeanArbel) [inline Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/README.md:11]: Can you please add to the readme the standard documentation by running `demisto-sdk generate-docs` command? You should run it with a `.txt` file containing command executions examples.

see `demisto-sdk generate-docs -h` for more help.

File: Packs/RSTCloud/Playbooks/Domain_Enrichment_-_RST_Threat_Feed.yml [hunk 12]
```diff
@@ -0,0 +1,243 @@
+id: Domain Enrichment - RST Threat Feed
+version: -1
+fromversion: 5.5.0
```
Comment (DeanArbel) [inline Packs/RSTCloud/Playbooks/Domain_Enrichment_-_RST_Threat_Feed.yml:3]: Why is this fromversion `5.5.0` while the integration is 5.0.0?

File: Packs/RSTCloud/Playbooks/IP_Enrichment_-_External_-_RST_Threat_Feed.yml [hunk 13]
```diff
@@ -0,0 +1,554 @@
+id: IP Enrichment - External - RST Threat Feed
+version: -1
+fromversion: 5.5.0
```
Comment (DeanArbel) [inline Packs/RSTCloud/Playbooks/IP_Enrichment_-_External_-_RST_Threat_Feed.yml:3]: Why is this fromversion 5.5.0 while the integration is 5.0.0?

File: Packs/RSTCloud/Playbooks/URL_Enrichment_-_RST_Threat_Feed.yml [hunk 14]
```diff
@@ -0,0 +1,607 @@
+id: URL Enrichment - RST Threat Feed
+version: -1
+fromversion: 5.5.0
```
Comment (DeanArbel) [inline Packs/RSTCloud/Playbooks/URL_Enrichment_-_RST_Threat_Feed.yml:3]: Why is this fromversion 5.5.0 while the integration is 5.0.0?

**Review by DeanArbel — CHANGES_REQUESTED** [review]
Very good work, this pack will be a nice addition to XSOAR.

* Please note the Logo should match the requirements (120x50 px PNG, transparent background, <10kB in Size) - your Logo is a different size.
* See the rest of my comments below.

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+    }
+
+    command = demisto.command()
+    LOG(f'RST: Command being called is {command}')
+
+    try:
+        if command == 'test-module':
+            demisto.results(test_module(client))
+        elif command in commands:
+            return_outputs(*commands[command](client, demisto.args()))
```
[the remark this answers is not in the document]
Comment (k1r10n) [reply Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py]: done. using CommandResults now

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py [hunk 16]
```diff
@@ -0,0 +1,429 @@
...
+    return markdown, context, results
+
+
+''' MAIN FUNCTION '''
+
+
+def main():
+    params = demisto.params()
+    client = Client(params.get('apikey'), params.get('url'), params.get('insecure', False), params.get('proxy', False))
+    LOG('RST: Client initialised...')
```
[the remark this answers is not in the document]
Comment (k1r10n) [reply Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py]: removed. was used for debugging

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py (hunk 1 above)
Comment (k1r10n) [reply Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/RSTCloudThreatFeedAPI.py]: updated

File: Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/README.md (hunk 11 above)
Comment (k1r10n) [reply Packs/RSTCloud/Integrations/RSTCloudThreatFeedAPI/README.md]: thanks. updated.

File: Packs/RSTCloud/Playbooks/Domain_Enrichment_-_RST_Threat_Feed.yml (hunk 12 above)
Comment (k1r10n) [reply Packs/RSTCloud/Playbooks/Domain_Enrichment_-_RST_Threat_Feed.yml]: updated to be 5.0.0

File: Packs/RSTCloud/Playbooks/IP_Enrichment_-_External_-_RST_Threat_Feed.yml (hunk 13 above)
Comment (k1r10n) [reply Packs/RSTCloud/Playbooks/IP_Enrichment_-_External_-_RST_Threat_Feed.yml]: updated to be 5.0.0

File: Packs/RSTCloud/Playbooks/URL_Enrichment_-_RST_Threat_Feed.yml (hunk 14 above)
Comment (k1r10n) [reply Packs/RSTCloud/Playbooks/URL_Enrichment_-_RST_Threat_Feed.yml]: updated to be 5.0.0

**Review by DeanArbel — APPROVED** [review]
Great job!

**Pull request merged**